ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala

```diff
@@ -58,6 +59,44 @@ final class SemanticTests extends LedgerTestSuite {
       }
   })
 
+  test(
+    "SemanticConcurrentDoubleSpend",
+    "Cannot concurrently double spend across transactions",
+    allocate(TwoParties, SingleParty),
+  )(implicit ec => {
+    case Participants(Participant(alpha, payer, owner1), Participant(beta, owner2)) =>
+      // This test create a contract and then concurrently archives it several times
+      // through two different participants
+      val contracts = 10 // Number of contracts to create
+      val archives = 10 // Number of concurrent archives per contract
+      Future
+        .traverse((1 to contracts).toList)(c =>
+          for {
+            shared <- alpha.create(payer, SharedContract(payer, owner1, owner2))
+            _ <- synchronize(alpha, beta)
+            results <- Future.traverse((1 to archives).toList)(i =>
+              i % 2 match {
+                case 0 =>
+                  alpha
+                    .exercise(owner1, shared.exerciseSharedContract_Consume1)
+                    .transform(Success(_))
+                case 1 =>
+                  beta
+                    .exercise(owner2, shared.exerciseSharedContract_Consume2)
+                    .transform(Success(_))
+            })
```

Comment:
You could potentially make use of `Seq.fill` here to make this a bit more declarative, though I'm not sure whether it's bettor not. Something like:

```suggestion
            exercises = Seq.fill(archives / 2)(Seq(
              (alpha, owner1, shared.exerciseSharedContract_Consume1),
              (beta, owner2, shared.exerciseSharedContract_Consume2),
            )).flatten
            results <- Future.traverse(exercises) {
              case (participant, owner, consume) =>
                participant.exercise(owner, consume).transform(Success(_))
            }
```

_I_ prefer it, but you don't have to agree.